src/components/TextInput/TextInput.tsx

```diff
@@ -0,0 +1,266 @@
+import React, {useMemo, useRef, useState} from 'react';
+import {
+  Animated,
+  TextInput as InputField,
+  NativeSyntheticEvent,
+  StyleSheet,
+  Text,
+  TextInputFocusEventData,
+  View,
+} from 'react-native';
+import {fontStyles} from '../../constants';
+import {useThemeColor} from '../../hooks';
+import {Strings} from '../../i18n';
+import {horizontalScale, verticalScale} from '../../utils';
+import {Label} from '../Label';
+import {TextInputProps} from './TextInput.types';
+
+export const TextInput = (props: TextInputProps) => {
+  const {
+    mode = 'default',
+    onChangeText,
+    hint,
+    style,
+    placeholder,
+    hintStyle,
+    error,
+    errorMessage,
+    errorMessageStyle,
+    label,
+    labelStyle,
+    editable,
+    inputStyle,
+    left,
+    right,
+    value,
+    ...restProps
+  } = props;
+  const [focused, setFocused] = useState(false);
+  const {colors} = useThemeColor();
+
+  const getStyle = useMemo(() => {
+    if (mode === 'outline') {
+      return styles.outline;
+    } else if (mode === 'border-less') {
+      return styles.borderLess;
+    }
+    return;
+  }, [mode]);
+
+  const getDisabledStyle = useMemo(() => {
+    if (!editable) {
+      if (mode === 'outline') {
+        return [
+          styles.outlineDisable,
+          {
+            borderColor: colors.placeholder,
+            backgroundColor: colors.disabled,
+          },
+        ];
+      } else if (mode === 'default') {
+        return [styles.defaultDisable, {borderBottomColor: colors.placeholder}];
+      }
+    } else {
+      return;
+    }
+  }, [colors.disabled, colors.placeholder, editable, mode]);
+
+  const floatingLabel = useRef(new Animated.Value(0)).current;
+
+  const yInterpolate = floatingLabel.interpolate({
+    inputRange: [0, 1],
+    outputRange: [4, -15],
+  });
+
+  const animation = {
+    transform: [
+      {
+        translateY: yInterpolate,
+      },
+    ],
+  };
+
+  const moveLabelTop = () => {
+    Animated.timing(floatingLabel, {
+      toValue: 1,
+      duration: 200,
+      useNativeDriver: true,
+    }).start();
+  };
+
+  const moveLabelDown = () => {
+    Animated.timing(floatingLabel, {
+      toValue: 0,
+      duration: 200,
+      useNativeDriver: true,
+    }).start();
+  };
+
+  const onFocusHandler = () => {
+    moveLabelTop();
+  };
+
+  const onBlurHandler = (e: string) => {
+    if (e?.length === 0 || value?.length === 0) {
+      moveLabelDown();
+    }
+  };
+
+  const renderAnimatedLabel = () => {
+    const animatedLabelStyle = [
+      animation,
+      styles.animatedLabel,
+      mode === 'outline'
+        ? {
+            backgroundColor: colors.background,
+          }
+        : null,
+    ];
+
+    const renderLabelStyle = [
```

Comment:
It would be better to keep animatedLabelStyle and renderLabelStyle outside renderAnimatedLabel and use memo